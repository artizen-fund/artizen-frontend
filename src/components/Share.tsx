import { useContext, useState } from 'react'
import styled from 'styled-components'
import { CopyToClipboard } from 'react-copy-to-clipboard'
import { Button, CloseButton, Icon } from '@components'
import { rgba, LayoutContext, titleCase } from '@lib'
import { palette, breakpoint, typography } from '@theme'
import { sharing } from '@copy/common'

interface ShareModalAttrs {
  mode?: 'home' | 'project' | 'postTransaction'
  destination?: string
  projectTitle?: string
  projectMember?: string
  twitterHandle?: string
  artizenHandle?: string
}

const ShareTransactionModal = () => {
  const { toggleModal } = useContext(LayoutContext)

  // note: this all shared with Share.tsx
  // we should DRY and abstract it
  const { modalAttrs } = useContext(LayoutContext)
  const { mode, destination, projectTitle, artizenHandle, twitterHandle } = modalAttrs as ShareModalAttrs

  const { modalTitle, modalDescription, shareCopy } = sharing[mode || 'home']

  const projectCreator = !!twitterHandle ? `@${twitterHandle}` : !!artizenHandle ? artizenHandle : 'the creator'

  console.log('projectCreator  ', projectCreator)

  const link = `https://artizen.fund${!!destination ? destination : ''}`

  const parsedShareCopy = shareCopy
    .replace('SHARE_LINK', link)
    .replace('PROJECT_TITLE', titleCase(projectTitle))
    .replace('PROJECT_CREATOR', titleCase(projectCreator) || 'Member')
  const title = 'Artizen'
  const twitterLink = `https://twitter.com/intent/tweet?text=${encodeURIComponent(parsedShareCopy)}`
  const facebookLink = `https://www.facebook.com/sharer/sharer.php?u=${encodeURIComponent(link)}&t=${encodeURIComponent(
    parsedShareCopy,
  )}`
  const redditLink = `http://www.reddit.com/submit?url=${encodeURIComponent(
    parsedShareCopy,
  )}&title=${encodeURIComponent('Artizen Fund')}`
  const mailLink = encodeURIComponent(link)

  const [copied, setCopied] = useState(false)

  return (
    <Wrapper>
      {mode === 'postTransaction' && (
        <>
          <Icon
            glyph="tick"
            outline
            level={2}
            color="algae"
            label="Thank you for supporting this project. A receipt will be sent to your e-mail."
          />
          <Row>
            <Rule />
          </Row>
        </>
      )}

      <Copy>
        <Header>{modalTitle}</Header>
        <Subheader>{modalDescription}</Subheader>
      </Copy>
      <Row>
        <ButtonWithLabel color="moon">
          <StyledButton
            {...{ link, title }}
            onClick={() => window?.open(twitterLink, '_blank')}
            glyph="twitter"
            glyphOnly
          >
            Twitter
          </StyledButton>
          <ButtonLabel>Twitter</ButtonLabel>
        </ButtonWithLabel>
        <ButtonWithLabel color="moon">
          <StyledButton
            {...{ link, title }}
            onClick={() => window?.open(facebookLink, '_blank')}
            glyph="facebook"
            glyphOnly
          >
            Facebook
          </StyledButton>
          <ButtonLabel>Facebook</ButtonLabel>
        </ButtonWithLabel>
        <ButtonWithLabel color="moon">
          <StyledButton {...{ link, title }} onClick={() => window?.open(redditLink, '_blank')} glyph="share" glyphOnly>
            Reddit
          </StyledButton>
          <ButtonLabel>Reddit</ButtonLabel>
        </ButtonWithLabel>
      </Row>
      <Row>
        <Rule />
      </Row>
      <Buttons>
        <Button level={1} stretch onClick={() => (location.href = `mailto:?subject=${title}&body=${mailLink}`)}>
          Email Link
        </Button>
        <CopyToClipboard text={link} onCopy={() => setCopied(true)}>
          <Button level={1} stretch outline>
            {copied && <span>Copied!</span>}
            {!copied && <span>Copy Site URL</span>}
          </Button>
        </CopyToClipboard>
      </Buttons>
      <CloseButton onClick={() => toggleModal()} />
    </Wrapper>
  )
}

const Wrapper = styled.div`
  position: relative;
  width: 100%;
  display: flex;
  flex-direction: column;
  max-width: 500px;
  gap: 30px;

  padding: 20px;
  @media only screen and (min-width: ${breakpoint.tablet}px) {
    width: 416px;
    padding: 30px;
  }
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    width: 568px;
    padding: 40px;
  }
  @media only screen and (min-width: ${breakpoint.desktop}px) {
    width: 680px;
  }

  color: ${rgba(palette.night)};
  background-color: ${rgba(palette.white)};
  box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.12);
  @media (prefers-color-scheme: dark) {
    color: ${rgba(palette.moon)};
    background-color: ${rgba(palette.slate)};
    box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.4);
  }
`

const Copy = styled.div`
  display: flex;
  flex-direction: column;
  gap: 5px;
`

const Header = styled.header`
  ${typography.title.l3}
`

const Subheader = styled.p`
  ${typography.body.l2}
`

const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  gap: 45px;
`

const Buttons = styled(props => <Row {...props} />)`
  gap: 10px;
`

const StyledButton = styled(props => <Button {...props} />)`
  transform: translate3d(0, 0, 0) scale3d(1, 1, 1);
  transition: background-color 0.25s ease-in-out, transform 0.4s cubic-bezier(0.42, 0.97, 0.52, 1.49);

  @media only screen and (min-width: ${breakpoint.tablet}px) {
    &:hover {
      background-color: ${rgba(palette.algae)};
      transform: translate3d(0, -4px, 0) scale3d(1.08, 1.08, 1);
      & * {
        background-color: ${rgba(palette.white)};
      }
    }
  }
`

const ButtonWithLabel = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 15px;
`

const ButtonLabel = styled.div`
  ${typography.title.l4}
  color: ${rgba(palette.night)};
  @media (prefers-color-scheme: dark) {
    color: ${rgba(palette.moon)};
  }
`

const Rule = styled.hr`
  height: 1px;
  width: 100%;
  border: 0;
  outline: 0;
  background: ${rgba(palette.stone)};
  @media (prefers-color-scheme: dark) {
    background: ${rgba(palette.barracuda, 0.64)};
  }
`

export default ShareTransactionModal
