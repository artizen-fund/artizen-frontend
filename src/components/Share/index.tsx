import { useContext, useState } from 'react'
import styled from 'styled-components'
import { CopyToClipboard } from 'react-copy-to-clipboard'
import { Button } from '@components'
import { rgba, LayoutContext } from '@lib'
import { palette, breakpoint, typography } from '@theme'

const Share = () => {
  const { visibleModal, toggleModal } = useContext(LayoutContext)

  const link = 'https://artizen.fund'
  const title = 'Artizen'
  const twitterLink = `https://twitter.com/intent/tweet?text=${encodeURIComponent(title)}&url=${encodeURIComponent(
    link,
  )}`
  const facebookLink = `https://www.facebook.com/sharer/sharer.php?u=${encodeURIComponent(
    title,
  )}%20${encodeURIComponent(link)}`
  const redditLink = `http://www.reddit.com/submit?url=${encodeURIComponent(link)}&title=${encodeURIComponent(title)}`
  const mailLink = encodeURIComponent(link)

  const [copied, setCopied] = useState(false)

  const visible = visibleModal === 'share'

  return (
    <Wrapper>
      <Row>
        <Header color="white_always">Sharing is caring, give back with a like, comment or share</Header>
      </Row>
      <Row>
        <ButtonWithLabel color="moon">
          <Button {...{ link, title }} onClick={() => window?.open(twitterLink, '_blank')} glyph="twitter" glyphOnly>
            Twitter
          </Button>
          <SubLabel>Twitter</SubLabel>
        </ButtonWithLabel>
        <ButtonWithLabel color="moon">
          <Button {...{ link, title }} onClick={() => window?.open(facebookLink, '_blank')} glyph="facebook" glyphOnly>
            Facebook
          </Button>
          <SubLabel>Facebook</SubLabel>
        </ButtonWithLabel>
        <ButtonWithLabel color="moon">
          <Button {...{ link, title }} onClick={() => window?.open(redditLink, '_blank')} glyph="share" glyphOnly>
            Reddit
          </Button>
          <SubLabel>Reddit</SubLabel>
        </ButtonWithLabel>
        <ButtonWithLabel color="moon">
          <Button
            {...{ link, title }}
            onClick={() => (location.href = `mailto:?subject=${title}&body=${mailLink}`)}
            glyph="plane"
            glyphOnly
          >
            Plane
          </Button>
          <SubLabel>Plane</SubLabel>
        </ButtonWithLabel>
      </Row>
      <Row>
        <CopyToClipboard text={link} onCopy={() => setCopied(true)}>
          <Button level={1} stretch>
            {copied && <span>Copied!</span>}
            {!copied && <span>Copy Site URL</span>}
          </Button>
        </CopyToClipboard>
      </Row>
    </Wrapper>
  )
}

const Wrapper = styled.div`
  position: relative;
  width: 100%;
  display: flex;
  flex-direction: column;
  max-width: 500px;
  gap: 30px;

  padding: 20px;
  @media only screen and (min-width: ${breakpoint.tablet}px) {
    padding: 30px;
  }
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    padding: 40px;
  }

  background-color: ${rgba(palette.night)};
  box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.12);
  @media (prefers-color-scheme: dark) {
    background-color: ${rgba(palette.slate)};
    box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.4);
  }
  border-radius: 16px;
`

const Header = styled.header`
  ${typography.title.l3}
  color: ${rgba(palette.moon)};
`

const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`

const ButtonWithLabel = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 15px;
`

const SubLabel = styled.div`
  ${typography.label.l2}
  color: ${rgba(palette.barracuda)};
`

export default Share
