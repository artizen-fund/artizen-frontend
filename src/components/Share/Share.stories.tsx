import Share from './'

export default {
  title: 'components/Share',
  component: Share,
  argTypes: {},
}

export const ShareComponent = (props: any) => <Share {...props} />
