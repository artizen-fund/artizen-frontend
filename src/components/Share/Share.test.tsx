import renderer from 'react-test-renderer'
import Share from './'

describe('Share', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Share />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
