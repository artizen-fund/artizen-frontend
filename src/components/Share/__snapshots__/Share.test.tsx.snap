// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share renders correctly 1`] = `
<div
  className="sc-d5b3c762-0 gbKcTJ"
>
  <div
    className="sc-d5b3c762-1 cFOooI"
  >
    <header
      className="sc-d5b3c762-2 kWwRtU"
    >
      Sharing is caring
    </header>
    <p
      className="sc-d5b3c762-3 dXoovD"
    >
      Help give back to the Artizen community by sharing this page with a friend.
    </p>
  </div>
  <div
    className="sc-d5b3c762-4 idlffj"
  >
    <div
      className="sc-d5b3c762-7 kuGJIt"
      color="moon"
    >
      <button
        className="sc-313ded62-2 bYllqi sc-d5b3c762-6 gHdsaS "
        onClick={[Function]}
        title="Artizen"
      >
        <div
          className="sc-3ba9b100-0 kSqsgh sc-313ded62-0 ggGmkW"
        />
        <span>
          Twitter
        </span>
      </button>
      <div
        className="sc-d5b3c762-8 fANDRo"
      >
        Twitter
      </div>
    </div>
    <div
      className="sc-d5b3c762-7 kuGJIt"
      color="moon"
    >
      <button
        className="sc-313ded62-2 bYllqi sc-d5b3c762-6 gHdsaS "
        onClick={[Function]}
        title="Artizen"
      >
        <div
          className="sc-3ba9b100-0 qXlHM sc-313ded62-0 ggGmkW"
        />
        <span>
          Facebook
        </span>
      </button>
      <div
        className="sc-d5b3c762-8 fANDRo"
      >
        Facebook
      </div>
    </div>
    <div
      className="sc-d5b3c762-7 kuGJIt"
      color="moon"
    >
      <button
        className="sc-313ded62-2 bYllqi sc-d5b3c762-6 gHdsaS "
        onClick={[Function]}
        title="Artizen"
      >
        <div
          className="sc-3ba9b100-0 iZJTeV sc-313ded62-0 ggGmkW"
        />
        <span>
          Reddit
        </span>
      </button>
      <div
        className="sc-d5b3c762-8 fANDRo"
      >
        Reddit
      </div>
    </div>
  </div>
  <div
    className="sc-d5b3c762-4 idlffj"
  >
    <hr
      className="sc-d5b3c762-9 blXaET"
    />
  </div>
  <div
    className="sc-d5b3c762-4 idlffj sc-d5b3c762-5 koKcnQ"
  >
    <button
      className="sc-313ded62-2 jCRIin undefined "
      onClick={[Function]}
    >
      <span>
        Email Link
      </span>
    </button>
    <button
      className="sc-313ded62-2 jzVwCg undefined "
      onClick={[Function]}
    >
      <span>
        <span>
          Copy Site URL
        </span>
      </span>
    </button>
  </div>
</div>
`;
